engine/execution/ingestion/throttle.go

```diff
@@ -0,0 +1,232 @@
+package ingestion
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+// CatchUpThreshold is the number of blocks that if the execution is far behind
+// the finalization then we will only lazy load the next unexecuted finalized
+// blocks until the execution has caught up
+const CatchUpThreshold = 500
+
+// BlockThrottle is a helper struct that throttles the unexecuted blocks to be sent
+// to the block queue for execution.
+// It is useful for case when execution is falling far behind the finalization, in which case
+// we want to throttle the blocks to be sent to the block queue for fetching data to execute
+// them. Without throttle, the block queue will be flooded with blocks, and the network
+// will be flooded with requests fetching collections, and the EN might quickly run out of memory.
+type BlockThrottle struct {
+	// config
+	threshold int // catch up threshold
+
+	// state
+	mu        sync.Mutex
+	executed  uint64
+	finalized uint64
+	inited    bool
+
+	// notifier
+	processables chan<- flow.Identifier
+
+	// dependencies
+	log     zerolog.Logger
+	state   protocol.State
+	headers storage.Headers
+}
+
+func NewBlockThrottle(
+	log zerolog.Logger,
+	state protocol.State,
+	execState state.ExecutionState,
+	headers storage.Headers,
+	catchupThreshold int,
+) (*BlockThrottle, error) {
+	finalizedHead, err := state.Final().Head()
+	if err != nil {
+		return nil, fmt.Errorf("could not get finalized head: %w", err)
+	}
+
+	finalized := finalizedHead.Height
+	// TODO: implement GetHighestFinalizedExecuted for execution state when storehouse
```

Comment:
does this need to happen before the throttle is used or will it only be used with storehouse?

Reply:
The `GetHighestFinalizedExecuted`, which the throttle depends on, was only needed by the storehouse. The old ingestion engine does not use it, so there was no implementation for the case when storehouse is not used. 

Why `throttle` depends on `GetHighestFinalizedExecuted`?

Because the `throttle` needs to keep track of latest executed height and latest finalized height in order to determine whether execution is falling far behind, and later whether the execution has caught up close enough to the last finalized blocks. And these two state variables need to be initialized regardless using storehouse or not. 

For now, this is a reminder, but [I have implemented it in a separate branch](https://github.com/onflow/flow-go/blob/leo/ingestion-new-engine-working/engine/execution/state/state.go#L516), which I will make PR for it later.